File: NinePSharp/messages/Tsetattr.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Tsetattr : IMessage
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Tsetattr;
    public ushort Tag { get; }
    public uint Fid { get; }
    public uint Valid { get; }
    public uint Mode { get; }
    public uint Uid { get; }
    public uint Gid { get; }
    public ulong FileSize { get; }
    public ulong AtimeSec { get; }
    public ulong AtimeNsec { get; }
    public ulong MtimeSec { get; }
    public ulong MtimeNsec { get; }

    public Tsetattr(uint size, ushort tag, uint fid, uint valid, uint mode, uint uid, uint gid, ulong fileSize, ulong atimeSec, ulong atimeNsec, ulong mtimeSec, ulong mtimeNsec)
    {
        Size = size;
        Tag = tag;
        Fid = fid;
        Valid = valid;
        Mode = mode;
        Uid = uid;
        Gid = gid;
        FileSize = fileSize;
        AtimeSec = atimeSec;
        AtimeNsec = atimeNsec;
        MtimeSec = mtimeSec;
        MtimeNsec = mtimeNsec;
    }

    public Tsetattr(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        Fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        Valid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        Mode = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        Uid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        Gid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        FileSize = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8)); offset += 8;
        AtimeSec = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8)); offset += 8;
        AtimeNsec = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8)); offset += 8;
        MtimeSec = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8)); offset += 8;
        MtimeNsec = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8));
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Tsetattr);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Fid); offset += 4;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Valid); offset += 4;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Mode); offset += 4;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Uid); offset += 4;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Gid); offset += 4;
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), FileSize); offset += 8;
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), AtimeSec); offset += 8;
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), AtimeNsec); offset += 8;
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), MtimeSec); offset += 8;
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), MtimeNsec);
    }
}

File: NinePSharp.Messages.FSharp/messages/Tsetattr.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Tsetattr =
    { Size: uint32
      Tag: uint16
      Fid: uint32
      Valid: uint32
      Mode: uint32
      Uid: uint32
      Gid: uint32
      FileSize: uint64
      AtimeSec: uint64
      AtimeNsec: uint64
      MtimeSec: uint64
      MtimeNsec: uint64 }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Tsetattr
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Tsetattr span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Fid); offset <- offset + 4
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Valid); offset <- offset + 4
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Mode); offset <- offset + 4
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Uid); offset <- offset + 4
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Gid); offset <- offset + 4
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.FileSize); offset <- offset + 8
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.AtimeSec); offset <- offset + 8
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.AtimeNsec); offset <- offset + 8
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.MtimeSec); offset <- offset + 8
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.MtimeNsec)

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let valid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let mode = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let uid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let gid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let fileSize = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        offset <- offset + 8
        let atimeSec = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        offset <- offset + 8
        let atimeNsec = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        offset <- offset + 8
        let mtimeSec = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        offset <- offset + 8
        let mtimeNsec = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        { Size = size; Tag = tag; Fid = fid; Valid = valid; Mode = mode; Uid = uid; Gid = gid; FileSize = fileSize; AtimeSec = atimeSec; AtimeNsec = atimeNsec; MtimeSec = mtimeSec; MtimeNsec = mtimeNsec }

